client/asset/dcr/spv.go

```diff
@@ -0,0 +1,1009 @@
+	if err != nil {
+		return 0, err
+	}
+	const numStamp = 11
+	timestamps := make([]int64, 0, numStamp)
+	for {
+		timestamps = append(timestamps, iBlkHeader.Timestamp.Unix())
+		if iBlkHeader.Height == 0 || len(timestamps) == numStamp {
+			break
+		}
+		iBlkHeader, err = w.dcrWallet.BlockHeader(ctx, &iBlkHeader.PrevBlock)
+		if err != nil {
+			return 0, fmt.Errorf("info not found for previous block: %v", err)
+		}
+	}
+	sort.Slice(timestamps, func(i, j int) bool {
+		return timestamps[i] < timestamps[j]
+	})
+	return timestamps[len(timestamps)/2], nil
+}
+
+// GetBlock returns the MsgBlock.
+// Part of the Wallet interface.
+func (w *spvWallet) GetBlock(ctx context.Context, blockHash *chainhash.Hash) (*wire.MsgBlock, error) {
+	if block := w.cachedBlock(blockHash); block != nil {
+		return block, nil
+	}
+
+	blocks, err := w.spv.Blocks(ctx, []*chainhash.Hash{blockHash})
+	if err != nil {
+		return nil, err
+	}
+	if len(blocks) == 0 { // Shouldn't actually be possible.
+		return nil, fmt.Errorf("network returned 0 blocks")
+	}
+
+	block := blocks[0]
+	w.cacheBlock(block)
+	return block, nil
+}
+
+// GetTransaction returns the details of a wallet tx, if the wallet contains a
+// tx with the provided hash. Returns asset.CoinNotFoundError if the tx is not
+// found in the wallet.
+// Part of the Wallet interface.
+func (w *spvWallet) GetTransaction(ctx context.Context, txHash *chainhash.Hash) (*walletjson.GetTransactionResult, error) {
+	// copy-pasted from dcrwallet/internal/rpc/jsonrpc/methods.go
+	txd, err := w.dcrWallet.TxDetails(ctx, txHash)
+	if errors.Is(err, walleterrors.NotExist) {
+		return nil, asset.CoinNotFoundError
+	} else if err != nil {
+		return nil, err
+	}
+
+	_, tipHeight := w.MainChainTip(ctx)
+
+	var b strings.Builder
+	b.Grow(2 * txd.MsgTx.SerializeSize())
+	err = txd.MsgTx.Serialize(hex.NewEncoder(&b))
+	if err != nil {
+		return nil, err
+	}
+
+	ret := walletjson.GetTransactionResult{
+		TxID:            txHash.String(),
+		Hex:             b.String(),
+		Time:            txd.Received.Unix(),
+		TimeReceived:    txd.Received.Unix(),
+		WalletConflicts: []string{}, // Not saved
+	}
+
+	if txd.Block.Height != -1 {
+		ret.BlockHash = txd.Block.Hash.String()
+		ret.BlockTime = txd.Block.Time.Unix()
+		ret.Confirmations = int64(tipHeight - txd.Block.Height + 1)
+	}
+
+	var (
+		debitTotal  dcrutil.Amount
+		creditTotal dcrutil.Amount
+		fee         dcrutil.Amount
+		negFeeF64   float64
+	)
+	for _, deb := range txd.Debits {
+		debitTotal += deb.Amount
+	}
+	for _, cred := range txd.Credits {
+		creditTotal += cred.Amount
+	}
+	// Fee can only be determined if every input is a debit.
+	if len(txd.Debits) == len(txd.MsgTx.TxIn) {
+		var outputTotal dcrutil.Amount
+		for _, output := range txd.MsgTx.TxOut {
+			outputTotal += dcrutil.Amount(output.Value)
+		}
+		fee = debitTotal - outputTotal
+		negFeeF64 = (-fee).ToCoin()
+	}
+	ret.Amount = (creditTotal - debitTotal).ToCoin()
+	ret.Fee = negFeeF64
```

Comment:
**TLDR**: I'm generally uneasy about computing anything reliably from the `Credits` and `Debits` slices (namely `Fee` or `Amount`), but fortunately the full `GetTransactionResult` is mostly unsued, so can we consider paring down the interface to make it both easier to satisfy and more reliable?

----

Lately I have concluded we should avoid relying on both the `Credits` and `Debits` slices from `TxDetails` (or the `Amount` and `Fee` fields that we compute from them or get from `gettransaction` on the RPC side) because they are very often incomplete or incorrect.  (Both BTC and DCR)

As you noted above, `Fee can only be determined if every input is a debit`, and this is indeed trouble with mix transactions, just as an example.  Inspecting one of those in my wallet, the `"fee"` field is omitted, and `"amount"` is something negative like `"amount": -0.002399` which is presumably indicative of the net change to my wallet's balance.
The help for `gettransaction` hints at more general conditions `"fee": n.nnn,                     (numeric)         The total input value minus the total output value, or 0 if 'txid' is not a sent transaction`.  Mixed transactions are just one instance.

I chatted with @martonp about this here: https://github.com/decred/dcrdex/pull/1555#discussion_r871745203

To be clear, I don't think the issue is what we are doing with the output from `dcrWallet.TxDetails`, but that it has weak assurances about what it will include in `Credits` and `Debits`, plus the conditions are different between unconfirmed and confirmed txns.  In both `unminedTxDetails` and `minedTxDetails`, it just grabs what credits and debits it has, and looking through real wallet transactions they are often missing.

Anyway, for the purpose of tx fee, Marton did it the hard but reliable way by pulling prevout tnxs for the input values, and in the Electrum PR I've [pared down the `GetTransactionResult` type](https://github.com/decred/dcrdex/commit/2e857bdac9f61721a05e854b0a12efba924c4133#diff-8b451093a1f3c33c83eee1f754caa0125c399e7764c35ec57eb6b57a254be39d) because it's neigh impossible to completely satisfy.  I think this is OK since consumers generally just want (1) confs, (2) block info, and (3) raw tx hex.  Scanning through this PR, I think that's still the case.